src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveData.cs

```diff
@@ -25,13 +25,24 @@ internal class CodeActionResolveData
         /// </remarks>
         public string UniqueIdentifier { get; }
 
+        /// <summary>
+        /// Identifies the Code Action Provider which produced this Code Action.
+        /// </summary>
+        /// <remarks>
+        /// The unique identifier is currently set as:
+        /// name of top level code action provider + '|' + name of code action provider implementation
+        /// e.g. 'Add Await Code Action Provider | GetTitleWithConfigureAwait'
+        /// </remarks>
+        public string ProviderName { get; }
```

Comment:
Needs to be added to published API.

Reply:
Ah, I don't think we need the `ProviderName` property in this file actually. The properties in `CodeActionResolveData` are just used to carry over necessary information from our `CodeActionHandler` class to our `CodeActionResolveHandler` class. Since `CodeActionHandler` is the class that tells LSP what code actions are available, I don't think we need to carry over the `ProviderName` information to the resolve handler, since all the resolve handler does is populate the code action edits.

Reply:
Wouldn't we need this property here to ensure the `CodeAction.Data` payload actually carries the provider name?

> The properties in CodeActionResolveData are just used to carry over necessary information from our CodeActionHandler class to our CodeActionResolveHandler class.

In-between Razor get's the `CodeActionResolveData` and that's where I was intending to inspect the payload to consume the ProviderName. Is there a better way to ensure razor get's the provider name?

Reply:
I might be misunderstanding how things work here. Could you explain what Razor is doing in between Roslyn's `CodeActionsHandler` and `CodeActionResolveHandler`?

My initial impression was that per our conversation with Miguel, Razor would provide Roslyn with a list ProviderNames through a custom property to Roslyn's `CodeActionsHandler`. Roslyn would then filter code actions based on these provider names, and only return back a filtered set of code actions. Later, Roslyn would get code action resolve requests for each of the filtered code actions, and then would populate these code actions with the edit data.
^Is this correct? Please feel free to correct me if not.

Reply:
> I might be misunderstanding how things work here. Could you explain what Razor is doing in between Roslyn's CodeActionsHandler and CodeActionResolveHandler?

1. Razor sends a `textDocument/codeAction` request to the Roslyn `CodeActionsHandler`.
2. Roslyn runs its analyzers & **[NEW in this PR]** annotates the response `CodeAction.CodeActionResolveData` with the `ProviderName` property.
3. Razor receives the response from #2, inspects the `CodeAction.CodeActionResolveData.ProviderName` to identify the code action, determine whether or not it is supported, and take the appropriate action to display it to the user in a razor compatible manner.
4. [Optional]: Should it be necessary to resolve the code action (ex. user selects it), then razor sends a `textDocument/codeActionResolve` request to Roslyn's `CodeActionResolveHandler` with the initial `CodeAction.CodeActionResolveData`.
5. Roslyn returns resolved edits, razor remaps those edits & displays them to the user.

> My initial impression was that per our conversation with Miguel, Razor would provide Roslyn with a list ProviderNames through a custom property to Roslyn's CodeActionsHandler. Roslyn would then filter code actions based on these provider names, and only return back a filtered set of code actions. Later, Roslyn would get code action resolve requests for each of the filtered code actions, and then would populate these code actions with the edit data.

This was definitely one of the proposals we were discussing, but we ended up deciding against it for now in the interest of simplicity and reduce coupling. This also required potential LSP Spec changes, whereas annotating the `CodeActionResolveData` payload does not as the spec just specifies an [`object Data;`](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#textDocument_codeAction). The goal of this PR was to take the approach steps detailed above. I hope that helps clarify things, let me know if there are any other questions :)

Reply:
That was a great explanation, thank you! In that case, yeah, I definitely think the data payload needs the provider name.